src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/MvcServicesProvider.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Hosting.Internal;
+using Microsoft.AspNetCore.Mvc.Internal;
+using Microsoft.AspNetCore.Mvc.Razor.Compilation;
+using Microsoft.AspNetCore.Mvc.Razor.Internal;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.ObjectPool;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design.Internal
+{
+    public class MvcServicesProvider
+    {
+        private const string ConfigureMvcMethod = "ConfigureMvc";
+        private readonly string _projectPath;
+        private readonly string _contentRoot;
+
+        public MvcServicesProvider(
+            string projectPath,
+            string contentRoot,
+            string configureCompilationType)
+        {
+            _projectPath = projectPath;
+            _contentRoot = contentRoot;
+
+            var configureCompilationAction = GetConfigureCompilationAction(configureCompilationType);
+            var serviceProvider = GetProvider(configureCompilationAction);
+
+            Host = serviceProvider.GetRequiredService<IMvcRazorHost>();
+            CompilationService = serviceProvider.GetRequiredService<ICompilationService>() as IRoslynCompilationService;
```

Comment:
This will change once we have the BuildManager code @cjqian will write. I'll remove the interface, but we'll leave the cast for now).